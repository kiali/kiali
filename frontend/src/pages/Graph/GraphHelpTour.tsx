import { PopoverPosition } from '@patternfly/react-core';
import { TourStopInfo, TourInfo } from 'components/Tour/TourStop';
import { GraphShortcuts } from './GraphToolbar/GraphShortcuts';
import { t } from 'utils/I18nUtils';
import { GraphShortcutsPF } from './GraphToolbar/GraphShortcutsPF';

export const GraphTourStops: { [name: string]: TourStopInfo } = {
  ContextualMenu: {
    name: t('Contextual Menu'),
    description: t(
      'Right-click a node or an edge to see the contextual menu with links to details, traffic and inbound/outbound metrics for the node or edge.'
    ),
    position: PopoverPosition.left,
    distance: 250
  },
  ContextualMenuPF: {
    name: t('Contextual Menu'),
    description: t(
      'Click kebab menu on a node label for links to details, traffic, inbound/outbound metrics and node graph.'
    ),
    position: PopoverPosition.left,
    distance: 250
  },
  Display: {
    name: t('Display'),
    description: t(
      'Set edge labeling, node badging, and various display options. Response-time edge labeling, security badging, and traffic animation may affect performance. Response-times reflect the 95th percentile.'
    ),
    position: PopoverPosition.rightStart
  },
  Find: {
    name: t('Find and Hide'),
    description: t(
      'Highlight or Hide graph elements via typed expressions. Click the dropdown for preset Find or Hide expressions. Click the Find/Hide help icon for details on the expression language.'
    ),
    position: PopoverPosition.bottom
  },
  Graph: {
    name: t('Graph'),
    description:
      "Click on a node or edge to see its summary and emphasize its end-to-end paths. Double-click a node to see a graph focused on that node.\nDouble-click an t('external namespace' node to navigate directly to the namespace in the node's text label. Shift-Drag to quickly zoom in.",
    position: PopoverPosition.left,
    distance: 250
  },
  GraphPF: {
    name: t('Graph'),
    description: t('Click on a node or edge to see its summary and emphasize its end-to-end paths.'),
    position: PopoverPosition.left,
    distance: 250
  },
  GraphTraffic: {
    name: t('Graph Traffic'),
    description: t(
      'Choose the traffic rates used to generate the graph. Each supported protocol offers one or more options. Unused protocols can be omitted.'
    ),
    position: PopoverPosition.bottom
  },
  GraphType: {
    name: t('Graph Type'),
    description: t(
      'Select a workload, service or application graph view. An application view can optionally be versioned and relies on app and version labeling. Workload and service graphs provide physical and logical views, respectively.'
    ),
    position: PopoverPosition.right
  },
  Layout: {
    name: t('Layout selection'),
    description: t(
      'Select the graph layout for the mesh. Different layouts work best with different meshes. Find the layout that works best. Other buttons here provide zoom and fit-to-screen options.'
    ),
    position: PopoverPosition.right
  },
  Legend: {
    name: t('Legend'),
    description: t('Display the legend to learn about what the different shapes, colors and backgrounds mean.'),
    position: PopoverPosition.rightEnd
  },
  Namespaces: {
    name: t('Namespaces'),
    description: t('Select the namespaces you want to see in the graph.'),
    position: PopoverPosition.bottomStart
  },
  Shortcuts: {
    name: t('Shortcuts'),
    htmlDescription: GraphShortcuts,
    position: PopoverPosition.leftStart
  },
  ShortcutsPF: {
    name: t('Shortcuts'),
    htmlDescription: GraphShortcutsPF,
    position: PopoverPosition.leftStart
  },
  SidePanel: {
    name: t('Side Panel'),
    description: t(
      'The Side Panel shows details about the currently selected node or edge, otherwise the whole graph.'
    ),
    position: PopoverPosition.left
  },
  TimeRange: {
    name: t('Time Range & Replay'),
    description: t(
      'Select how often to refresh the graph and how much historical metric data is used to build the graph, per refresh. For example "Last 5m" means use the most recent 5 minutes of request metric data.  To replay a historical time window click the replay icon.  This replaces the current time range controls with the replay toolbar.'
    ),
    position: PopoverPosition.bottomEnd
  }
};

export const GraphTour: TourInfo = {
  name: t('GraphTour'),
  stops: [
    GraphTourStops.Shortcuts,
    GraphTourStops.Namespaces,
    GraphTourStops.GraphTraffic,
    GraphTourStops.GraphType,
    GraphTourStops.Display,
    GraphTourStops.Find,
    GraphTourStops.TimeRange,
    GraphTourStops.Graph,
    GraphTourStops.ContextualMenu,
    GraphTourStops.SidePanel,
    GraphTourStops.Layout,
    GraphTourStops.Legend
  ]
};

export const GraphTourPF: TourInfo = {
  name: t('GraphTour'),
  stops: [
    GraphTourStops.ShortcutsPF,
    GraphTourStops.Namespaces,
    GraphTourStops.GraphTraffic,
    GraphTourStops.GraphType,
    GraphTourStops.Display,
    GraphTourStops.Find,
    GraphTourStops.TimeRange,
    GraphTourStops.GraphPF,
    GraphTourStops.ContextualMenuPF,
    GraphTourStops.SidePanel,
    GraphTourStops.Layout,
    GraphTourStops.Legend
  ]
};
