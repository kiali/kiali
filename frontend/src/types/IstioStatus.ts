import { t } from 'utils/I18nUtils';

export enum Status {
  Healthy = 'Healthy',
  Unhealthy = 'Unhealthy',
  Unreachable = 'Unreachable',
  NotFound = 'NotFound',
  NotReady = 'NotReady'
}

export const statusMsg = {
  [Status.Healthy]: t('Healthy'),
  [Status.NotFound]: t('Not found'),
  [Status.NotReady]: t('Not ready'),
  [Status.Unhealthy]: t('Not healthy'),
  [Status.Unreachable]: t('Unreachable')
};

export const statusSeverity: Record<Status, number> = {
  [Status.Unhealthy]: 5,
  [Status.Unreachable]: 4,
  [Status.NotReady]: 3,
  [Status.NotFound]: 2,
  [Status.Healthy]: 1
};

export interface ComponentStatus {
  cluster: string;
  is_core: boolean;
  name: string;
  status: Status;
}

export interface IstiodResourceThresholds {
  cpu: number;
  memory: number;
}
