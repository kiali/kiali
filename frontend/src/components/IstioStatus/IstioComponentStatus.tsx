import * as React from 'react';
import { ComponentStatus, Status, statusMsg } from '../../types/IstioStatus';
import { PFColors } from '../Pf/PfColors';
import {
  CheckCircleIcon,
  ExclamationCircleIcon,
  ExclamationTriangleIcon,
  MinusCircleIcon
} from '@patternfly/react-icons';
import { Split, SplitItem } from '@patternfly/react-core';
import { IconProps, createIcon } from 'config/KialiIcon';
import { kialiStyle } from 'styles/StyleUtils';
import { useKialiTranslation } from 'utils/I18nUtils';

type Props = {
  componentStatus: ComponentStatus;
};

const ErrorCoreComponent: IconProps = {
  color: PFColors.Danger,
  icon: ExclamationCircleIcon
};

const ErrorAddonComponent: IconProps = {
  color: PFColors.Warning,
  icon: ExclamationTriangleIcon
};

const NotReadyComponent: IconProps = {
  color: PFColors.Info,
  icon: MinusCircleIcon
};

const SuccessComponent: IconProps = {
  color: PFColors.Success,
  icon: CheckCircleIcon
};

// Mapping Valid-Core to Icon representation.
const validToIcon: { [valid: string]: IconProps } = {
  'false-false': ErrorAddonComponent,
  'false-true': ErrorCoreComponent,
  'true-false': SuccessComponent,
  'true-true': SuccessComponent
};

const splitItemStyle = kialiStyle({
  textAlign: 'left'
});

export const IstioComponentStatus: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const renderIcon = (status: Status, isCore: boolean): React.ReactNode => {
    let compIcon = validToIcon[`${status === Status.Healthy}-${isCore}`];

    if (status === Status.NotReady) {
      compIcon = NotReadyComponent;
    }

    compIcon.className = kialiStyle({
      marginTop: '0.25rem'
    });

    return createIcon(compIcon);
  };

  const renderCells = (): React.ReactNode => {
    const comp = props.componentStatus;

    return [
      <Split key={`cell-status-icon-${comp.name}`} hasGutter={true} className={splitItemStyle}>
        <SplitItem>{renderIcon(props.componentStatus.status, props.componentStatus.isCore)}</SplitItem>
        <SplitItem isFilled={true}>{comp.name}</SplitItem>
        <SplitItem>{t(statusMsg[comp.status])}</SplitItem>
      </Split>
    ];
  };

  return <>{renderCells()}</>;
};
