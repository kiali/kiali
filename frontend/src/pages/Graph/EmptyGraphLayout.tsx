import * as React from 'react';
import {
  Button,
  ButtonVariant,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  EmptyStateHeader,
  EmptyStateFooter
} from '@patternfly/react-core';
import { kialiStyle } from 'styles/StyleUtils';
import * as _ from 'lodash';
import { Namespace } from '../../types/Namespace';
import { KialiIcon } from '../../config/KialiIcon';
import { DecoratedGraphElements } from '../../types/Graph';
import { RefreshIntervalManual } from 'config/Config';
import { IntervalInMilliseconds } from 'types/Common';
import { t } from 'utils/I18nUtils';

type EmptyGraphLayoutProps = {
  action?: any;
  elements?: DecoratedGraphElements;
  error?: string;
  isError: boolean;
  isLoading?: boolean;
  isMiniGraph: boolean;
  loaded: boolean;
  namespaces?: Namespace[];
  refreshInterval?: IntervalInMilliseconds;
  showIdleNodes?: boolean;
  toggleIdleNodes?: () => void;
};

const emptyStateStyle = kialiStyle({
  height: '98%',
  marginRight: 'auto',
  marginLeft: 'auto',
  marginBottom: 10,
  marginTop: 10
});

type EmptyGraphLayoutState = {};

export class EmptyGraphLayout extends React.Component<EmptyGraphLayoutProps, EmptyGraphLayoutState> {
  shouldComponentUpdate(nextProps: EmptyGraphLayoutProps): boolean {
    const currentIsEmpty = this.props.elements === undefined || _.isEmpty(this.props.elements.nodes);
    const nextIsEmpty = nextProps.elements === undefined || _.isEmpty(nextProps.elements.nodes);

    // Update if we have elements and we are not loading
    if (!nextProps.isLoading && !nextIsEmpty) {
      return true;
    }

    // Update if we are going from having no elements to having elements or vice versa
    if (currentIsEmpty !== nextIsEmpty) {
      return true;
    }
    // Do not update if we have elements and the namespace didn't change, as this means we are refreshing
    return !(!nextIsEmpty && _.isEqual(this.props.namespaces, nextProps.namespaces));
  }

  namespacesText(): React.ReactElement | null {
    if (this.props.namespaces && this.props.namespaces.length > 0) {
      if (this.props.namespaces.length === 1) {
        return (
          <div>
            <b>{this.props.namespaces[0].name}</b>
          </div>
        );
      } else {
        const namespacesString = `${this.props.namespaces
          .slice(0, -1)
          .map(namespace => namespace.name)
          .join(',')} and ${this.props.namespaces[this.props.namespaces.length - 1].name}`;
        return (
          <div>
            <b>{namespacesString}</b>
          </div>
        );
      }
    }
    return null;
  }

  render(): React.ReactNode {
    if (this.props.isError) {
      return (
        <EmptyState id="empty-graph-error" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader
            titleText={t('Error loading graph')}
            icon={<EmptyStateIcon icon={KialiIcon.Error} />}
            headingLevel="h5"
          />
          <EmptyStateBody>{this.props.error}</EmptyStateBody>
        </EmptyState>
      );
    }

    if (this.props.isLoading) {
      return (
        <EmptyState id="empty-graph-is-loading" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader titleText={t('Loading graph')} headingLevel="h5" />
        </EmptyState>
      );
    }

    const isGraphEmpty = !this.props.elements || !this.props.elements.nodes || this.props.elements.nodes.length < 1;

    if (this.props.namespaces?.length === 0) {
      return (
        <EmptyState id="empty-graph-no-namespace" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader titleText={t('No namespace is selected')} headingLevel="h5" />
          <EmptyStateBody>
            {t('There is currently no namespace selected, please select one using the Namespace selector.')}
          </EmptyStateBody>
        </EmptyState>
      );
    }

    if (this.props.refreshInterval === RefreshIntervalManual && !this.props.loaded && !this.props.isMiniGraph) {
      return (
        <EmptyState
          id="empty-graph-manual"
          data-test="manual-refresh"
          variant={EmptyStateVariant.lg}
          className={emptyStateStyle}
        >
          <EmptyStateHeader titleText={t('Manual refresh required')} headingLevel="h5" />
          <EmptyStateBody>
            {t(
              'The refresh interval is set to "Manual". To render the graph, select your desired filters and options and then click the Refresh button. Or, if preferred, change the setting to the desired interval.'
            )}
          </EmptyStateBody>
        </EmptyState>
      );
    }

    if (isGraphEmpty && !this.props.isMiniGraph) {
      return (
        <EmptyState id="empty-graph" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader titleText={t('Empty Graph')} headingLevel="h5" />
          <EmptyStateBody>
            {t(
              'There is currently no graph available for the selected namespaces. This usually means the namespaces do not have measurable traffic for the selected time period:'
            )}
            {this.namespacesText()}
            <br />
            {this.props.showIdleNodes && (
              <>
                {t(
                  'You are currently displaying "Idle nodes". Send requests to the service mesh and click the Refresh button.'
                )}
              </>
            )}
            {!this.props.showIdleNodes && (
              <>{t('You can enable "Idle Nodes" to display nodes that have yet to see any request traffic.')}</>
            )}
            <p>
              <br />
              {t('To read the FAQ entry about why your graph may be empty, ')}
              <a href={'https://kiali.io/docs/faq/graph/#emptygraph'} target="_blank" rel="noreferrer">
                {t('click here.')}
              </a>
            </p>
          </EmptyStateBody>
          <EmptyStateFooter>
            <Button
              onClick={this.props.showIdleNodes ? this.props.action : this.props.toggleIdleNodes}
              variant={ButtonVariant.primary}
            >
              {(this.props.showIdleNodes && <>{t('Refresh')}</>) || <>{t('Display idle nodes')}</>}
            </Button>
          </EmptyStateFooter>
        </EmptyState>
      );
    }

    if (isGraphEmpty && this.props.isMiniGraph) {
      return (
        <EmptyState id="empty-mini-graph" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader titleText={t('Empty Graph')} headingLevel="h5" />
          <EmptyStateBody>{t('No graph traffic for the time period.')}</EmptyStateBody>
        </EmptyState>
      );
    }

    return this.props.children;
  }
}
