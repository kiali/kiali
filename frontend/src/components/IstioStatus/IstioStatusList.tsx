import * as React from 'react';
import { List } from '@patternfly/react-core';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { IstioComponentStatus } from './IstioComponentStatus';
import { kialiStyle } from 'styles/StyleUtils';
import { useKialiTranslation } from 'utils/I18nUtils';

type Props = {
  cluster: string;
  status: ComponentStatus[];
};

const listStyle = kialiStyle({
  paddingLeft: 0,
  marginTop: 0,
  marginLeft: 0
});

export const IstioStatusList: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const nonhealthyComponents = (): ComponentStatus[] => {
    return props.status.filter((c: ComponentStatus) => c.status !== Status.Healthy);
  };

  const coreComponentsStatus = (): ComponentStatus[] => {
    return nonhealthyComponents().filter((s: ComponentStatus) => s.is_core);
  };

  const addonComponentsStatus = (): ComponentStatus[] => {
    return nonhealthyComponents().filter((s: ComponentStatus) => !s.is_core);
  };

  const renderComponentList = (): React.ReactNode => {
    const groups = {
      core: coreComponentsStatus,
      addon: addonComponentsStatus
    };

    return ['core', 'addon'].map((group: string) => {
      return (
        <React.Fragment key={`status-${group}`}>
          {groups[group]().map((status: ComponentStatus) => {
            return <IstioComponentStatus key={`status-${group}-${status.name}`} componentStatus={status} />;
          })}
        </React.Fragment>
      );
    });
  };

  return (
    <List id="istio-status" aria-label={t('Istio Component List')} className={listStyle}>
      {renderComponentList()}
    </List>
  );
};
