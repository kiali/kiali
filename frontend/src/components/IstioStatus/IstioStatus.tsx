import * as React from 'react';
import { SVGIconProps } from '@patternfly/react-icons/dist/js/createIcon';
import * as API from '../../services/Api';
import * as AlertUtils from '../../utils/AlertUtils';
import { TimeInMilliseconds } from '../../types/Common';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { MessageType } from '../../types/MessageCenter';
import { Namespace } from '../../types/Namespace';
import { KialiAppState } from '../../store/Store';
import { istioStatusSelector, namespaceItemsSelector } from '../../store/Selectors';
import { bindActionCreators } from 'redux';
import { IstioStatusActions } from '../../actions/IstioStatusActions';
import { connect } from 'react-redux';
import { Tooltip, TooltipPosition } from '@patternfly/react-core';
import { IstioStatusList } from './IstioStatusList';
import { PFColors } from '../Pf/PfColors';
import { ResourcesFullIcon } from '@patternfly/react-icons';
import { KialiDispatch } from 'types/Redux';
import { NamespaceThunkActions } from '../../actions/NamespaceThunkActions';
import { connectRefresh } from '../Refresh/connectRefresh';
import { kialiStyle } from 'styles/StyleUtils';
import { IconProps, createIcon } from 'config/KialiIcon';
import { Link } from 'react-router-dom-v5-compat';
import { useKialiTranslation } from 'utils/I18nUtils';
import { MASTHEAD } from 'components/Nav/Masthead/Masthead';
import { isControlPlaneAccessible } from '../../utils/MeshUtils';

type ReduxStateProps = {
  namespaces?: Namespace[];
  status: ComponentStatus[];
};

type ReduxDispatchProps = {
  refreshNamespaces: () => void;
  setIstioStatus: (istioStatus: ComponentStatus[]) => void;
};

type StatusIcons = {
  ErrorIcon?: React.ComponentClass<SVGIconProps>;
  HealthyIcon?: React.ComponentClass<SVGIconProps>;
  InfoIcon?: React.ComponentClass<SVGIconProps>;
  WarningIcon?: React.ComponentClass<SVGIconProps>;
};

type Props = ReduxStateProps &
  ReduxDispatchProps & {
    cluster?: string;
    icons?: StatusIcons;
    lastRefreshAt: TimeInMilliseconds;
    location?: string;
  };

const ValidToColor = {
  'true-true-true': PFColors.Danger,
  'true-true-false': PFColors.Danger,
  'true-false-true': PFColors.Danger,
  'true-false-false': PFColors.Danger,
  'false-true-true': PFColors.Warning,
  'false-true-false': PFColors.Warning,
  'false-false-true': PFColors.Info,
  'false-false-false': PFColors.Success
};

const defaultIcons = {
  ErrorIcon: ResourcesFullIcon,
  HealthyIcon: ResourcesFullIcon,
  InfoIcon: ResourcesFullIcon,
  WarningIcon: ResourcesFullIcon
};

const iconStyle = kialiStyle({
  marginLeft: '2rem',
  fontSize: '1rem'
});

export const meshLinkStyle = kialiStyle({
  display: 'flex',
  justifyContent: 'center',
  marginTop: '0.75rem',
  $nest: {
    '& > span': {
      marginRight: '0.5rem'
    }
  }
});

export const IstioStatusComponent: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const { cluster, namespaces, setIstioStatus, refreshNamespaces, lastRefreshAt } = props;

  React.useEffect(() => {
    refreshNamespaces();
  }, [refreshNamespaces]);

  const fetchStatus = React.useCallback((): void => {
    API.getIstioStatus(cluster)
      .then(response => {
        setIstioStatus(response.data);
      })
      .catch(error => {
        // User without namespaces can't have access to mTLS information. Reduce severity to info.
        const informative = namespaces && namespaces.length < 1;

        if (informative) {
          AlertUtils.addError(t('Istio deployment status disabled.'), error, 'default', MessageType.INFO);
        } else {
          AlertUtils.addError(t('Error fetching Istio deployment status.'), error, 'default', MessageType.ERROR);
        }
      });
  }, [cluster, namespaces, setIstioStatus, t]);

  React.useEffect(() => {
    fetchStatus();
  }, [lastRefreshAt, fetchStatus]);

  const tooltipContent = (): React.ReactNode => {
    return (
      <>
        <IstioStatusList status={props.status} />

        {!props.location?.endsWith('/mesh') && isControlPlaneAccessible() && (
          <div className={meshLinkStyle}>
            <span>{t('More info at')}</span>
            <Link to="/mesh">{t('Mesh page')}</Link>
          </div>
        )}
      </>
    );
  };

  const tooltipColor = (): string => {
    let coreUnhealthy = false;
    let addonUnhealthy = false;
    let notReady = false;

    Object.keys(props.status ?? {}).forEach((compKey: string) => {
      const { status, is_core } = props.status[compKey];
      const isNotReady: boolean = status === Status.NotReady;
      const isUnhealthy: boolean = status !== Status.Healthy && !isNotReady;

      if (is_core) {
        coreUnhealthy = coreUnhealthy || isUnhealthy;
      } else {
        addonUnhealthy = addonUnhealthy || isUnhealthy;
      }

      notReady = notReady || isNotReady;
    });

    return ValidToColor[`${coreUnhealthy}-${addonUnhealthy}-${notReady}`];
  };

  const healthyComponents = (): boolean => {
    return props.status.reduce((healthy: boolean, compStatus: ComponentStatus) => {
      return healthy && compStatus.status === Status.Healthy;
    }, true);
  };

  if (!healthyComponents()) {
    const icons = props.icons ? { ...defaultIcons, ...props.icons } : defaultIcons;
    const iconColor = tooltipColor();
    let icon = ResourcesFullIcon;
    let dataTest = 'istio-status';

    if (iconColor === PFColors.Danger) {
      icon = icons.ErrorIcon;
      dataTest = `${dataTest}-danger`;
    } else if (iconColor === PFColors.Warning) {
      icon = icons.WarningIcon;
      dataTest = `${dataTest}-warning`;
    } else if (iconColor === PFColors.Info) {
      icon = icons.InfoIcon;
      dataTest = `${dataTest}-info`;
    } else if (iconColor === PFColors.Success) {
      icon = icons.HealthyIcon;
      dataTest = `${dataTest}-success`;
    }

    const iconProps: IconProps = {
      className: iconStyle,
      dataTest: dataTest
    };

    const tooltipPosition = props.location === MASTHEAD ? TooltipPosition.bottom : TooltipPosition.top;

    return (
      <Tooltip position={tooltipPosition} enableFlip={true} content={tooltipContent()} maxWidth="25rem">
        {createIcon(iconProps, icon, iconColor)}
      </Tooltip>
    );
  }

  return null;
};

const mapStateToProps = (state: KialiAppState): ReduxStateProps => ({
  status: istioStatusSelector(state),
  namespaces: namespaceItemsSelector(state)
});

const mapDispatchToProps = (dispatch: KialiDispatch): ReduxDispatchProps => ({
  setIstioStatus: bindActionCreators(IstioStatusActions.setinfo, dispatch),
  refreshNamespaces: () => {
    dispatch(NamespaceThunkActions.fetchNamespacesIfNeeded());
  }
});

export const IstioStatus = connectRefresh(connect(mapStateToProps, mapDispatchToProps)(IstioStatusComponent));
