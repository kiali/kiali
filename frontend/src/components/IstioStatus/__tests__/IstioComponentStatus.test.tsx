import * as React from 'react';
import { ComponentStatus, Status } from '../../../types/IstioStatus';
import { IstioComponentStatus } from '../IstioComponentStatus';
import { shallowToJson } from 'enzyme-to-json';
import { shallow, ShallowWrapper } from 'enzyme';
import { CLUSTER_DEFAULT } from '../../../types/Graph';

const mockComponent = (cs: ComponentStatus): ShallowWrapper => {
  return shallow(<IstioComponentStatus componentStatus={cs} />);
};

describe('IstioComponentStatus renders', () => {
  it('success icon when core component is running', () => {
    const wrapper = mockComponent({
      cluster: CLUSTER_DEFAULT,
      name: 'isito-ingress',
      status: Status.Healthy,
      is_core: true
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('error icon when core component is not running', () => {
    const wrapper = mockComponent({
      cluster: CLUSTER_DEFAULT,
      name: 'isito-ingress',
      status: Status.Unhealthy,
      is_core: true
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('error icon when core component is not found', () => {
    const wrapper = mockComponent({
      cluster: CLUSTER_DEFAULT,
      name: 'isito-ingress',
      status: Status.NotFound,
      is_core: true
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('success icon when core component is running', () => {
    const wrapper = mockComponent({
      cluster: CLUSTER_DEFAULT,
      name: 'prometheus',
      status: Status.Healthy,
      is_core: false
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('warning icon when core component is not running', () => {
    const wrapper = mockComponent({
      cluster: CLUSTER_DEFAULT,
      name: 'prometheus',
      status: Status.Unhealthy,
      is_core: false
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('warning icon when core component is not found', () => {
    const wrapper = mockComponent({
      cluster: CLUSTER_DEFAULT,
      name: 'prometheus',
      status: Status.NotFound,
      is_core: false
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('minus icon when core component is not ready', () => {
    const wrapper = mockComponent({
      cluster: CLUSTER_DEFAULT,
      name: 'core',
      status: Status.NotReady,
      is_core: true
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('minus icon when addon component is not ready', () => {
    const wrapper = mockComponent({
      cluster: CLUSTER_DEFAULT,
      name: 'addon',
      status: Status.NotReady,
      is_core: false
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
