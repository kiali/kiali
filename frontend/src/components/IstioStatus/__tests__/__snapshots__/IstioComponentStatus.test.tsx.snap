// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IstioComponentStatus renders error icon when core component is not found 1`] = `
<Fragment>
  <Split
    className="kiali_fuhkp1b"
    hasGutter={true}
    key="cell-status-icon-isito-ingress"
  >
    <SplitItem
      style={
        Object {
          "paddingLeft": "0.5rem",
        }
      }
    >
      <Icon
        className="kiali_fi2mdky kiali_f1h37jb8"
      >
        <ExclamationCircleIcon />
      </Icon>
    </SplitItem>
    <SplitItem
      isFilled={true}
    >
      isito-ingress
    </SplitItem>
    <SplitItem>
      Not found
    </SplitItem>
  </Split>
</Fragment>
`;

exports[`IstioComponentStatus renders error icon when core component is not running 1`] = `
<Fragment>
  <Split
    className="kiali_fuhkp1b"
    hasGutter={true}
    key="cell-status-icon-isito-ingress"
  >
    <SplitItem
      style={
        Object {
          "paddingLeft": "0.5rem",
        }
      }
    >
      <Icon
        className="kiali_fi2mdky kiali_f1h37jb8"
      >
        <ExclamationCircleIcon />
      </Icon>
    </SplitItem>
    <SplitItem
      isFilled={true}
    >
      isito-ingress
    </SplitItem>
    <SplitItem>
      Not healthy
    </SplitItem>
  </Split>
</Fragment>
`;

exports[`IstioComponentStatus renders minus icon when addon component is not ready 1`] = `
<Fragment>
  <Split
    className="kiali_fuhkp1b"
    hasGutter={true}
    key="cell-status-icon-addon"
  >
    <SplitItem
      style={
        Object {
          "paddingLeft": "0.5rem",
        }
      }
    >
      <Icon
        className="kiali_fi2mdky kiali_f1elceoh"
      >
        <MinusCircleIcon />
      </Icon>
    </SplitItem>
    <SplitItem
      isFilled={true}
    >
      addon
    </SplitItem>
    <SplitItem>
      Not ready
    </SplitItem>
  </Split>
</Fragment>
`;

exports[`IstioComponentStatus renders minus icon when core component is not ready 1`] = `
<Fragment>
  <Split
    className="kiali_fuhkp1b"
    hasGutter={true}
    key="cell-status-icon-core"
  >
    <SplitItem
      style={
        Object {
          "paddingLeft": "0.5rem",
        }
      }
    >
      <Icon
        className="kiali_fi2mdky kiali_f1elceoh"
      >
        <MinusCircleIcon />
      </Icon>
    </SplitItem>
    <SplitItem
      isFilled={true}
    >
      core
    </SplitItem>
    <SplitItem>
      Not ready
    </SplitItem>
  </Split>
</Fragment>
`;

exports[`IstioComponentStatus renders success icon when core component is running 1`] = `
<Fragment>
  <Split
    className="kiali_fuhkp1b"
    hasGutter={true}
    key="cell-status-icon-isito-ingress"
  >
    <SplitItem
      style={
        Object {
          "paddingLeft": "0.5rem",
        }
      }
    >
      <Icon
        className="kiali_fi2mdky kiali_f1r5cf3w"
      >
        <CheckCircleIcon />
      </Icon>
    </SplitItem>
    <SplitItem
      isFilled={true}
    >
      isito-ingress
    </SplitItem>
    <SplitItem>
      Healthy
    </SplitItem>
  </Split>
</Fragment>
`;

exports[`IstioComponentStatus renders success icon when core component is running 2`] = `
<Fragment>
  <Split
    className="kiali_fuhkp1b"
    hasGutter={true}
    key="cell-status-icon-prometheus"
  >
    <SplitItem
      style={
        Object {
          "paddingLeft": "0.5rem",
        }
      }
    >
      <Icon
        className="kiali_fi2mdky kiali_f1r5cf3w"
      >
        <CheckCircleIcon />
      </Icon>
    </SplitItem>
    <SplitItem
      isFilled={true}
    >
      prometheus
    </SplitItem>
    <SplitItem>
      Healthy
    </SplitItem>
  </Split>
</Fragment>
`;

exports[`IstioComponentStatus renders warning icon when core component is not found 1`] = `
<Fragment>
  <Split
    className="kiali_fuhkp1b"
    hasGutter={true}
    key="cell-status-icon-prometheus"
  >
    <SplitItem
      style={
        Object {
          "paddingLeft": "0.5rem",
        }
      }
    >
      <Icon
        className="kiali_fi2mdky kiali_frkkrth"
      >
        <ExclamationTriangleIcon />
      </Icon>
    </SplitItem>
    <SplitItem
      isFilled={true}
    >
      prometheus
    </SplitItem>
    <SplitItem>
      Not found
    </SplitItem>
  </Split>
</Fragment>
`;

exports[`IstioComponentStatus renders warning icon when core component is not running 1`] = `
<Fragment>
  <Split
    className="kiali_fuhkp1b"
    hasGutter={true}
    key="cell-status-icon-prometheus"
  >
    <SplitItem
      style={
        Object {
          "paddingLeft": "0.5rem",
        }
      }
    >
      <Icon
        className="kiali_fi2mdky kiali_frkkrth"
      >
        <ExclamationTriangleIcon />
      </Icon>
    </SplitItem>
    <SplitItem
      isFilled={true}
    >
      prometheus
    </SplitItem>
    <SplitItem>
      Not healthy
    </SplitItem>
  </Split>
</Fragment>
`;
